### ISSUE TITLE
Request parameter doesn't seem to be passed to form request object

### ISSUE THREAD

**Issue by TheLazzziest** [upgate/laravel-jsonrpc#6]
I found that you don't support form request object in laravel, because you haven't found request parameters initialization and passing them to form request or a default request object if it's passed to a controller action.
/upgate/laravel-jsonrpc/src/Server/RouteDispatcher.php
`if ($requestParams) {
            $params = $requestParams->getParams() ?: null;
            if ($params) {
                $areParamsNamed = $requestParams->areParamsNamed();
            }

        }

        $args = [];
        foreach ($method->getParameters() as $parameter) {
            $class = $parameter->getClass();
            if ($class) {
                try {
                    /** I suppose some initialization must be implemented here*/
                    $args[] = $this->container->make($class->name);
                } catch (ReflectionException $e) {
                    throw new InternalErrorException($e->getMessage(), 0, $e);
                }
            } else {
                if (null !== $params) {
                    if ($parameter->isVariadic()) {
                        foreach ($params as $key => $value) {
                            $args[] = $this->cast($value, $parameter);
                        }
                        break;
                    }

                    if ($areParamsNamed) {
                        $name = $parameter->getName();
                        if (array_key_exists($name, $params)) {
                            $args[] = $this->cast($params[$name], $parameter);
                            unset($params[$name]);
                            continue;
                        }
                    } else {
                        if (count($params)) {
                            $args[] = $this->cast(array_shift($params), $parameter);
                            continue;
                        }
                    }
                }
                try {
                    $args[] = $parameter->getDefaultValue();
                } catch (ReflectionException $e) {
                    throw new InvalidParamsException("\"{$parameter->getName()}\" is required", 0, $e);
                }
            }
        }`

**Comment by kbaryshnikov** [comment]
The reason why FormRequest is not supported is different.

FormRequest extends Illuminate\Http\Request, so it *is* an HTTP request, by its nature. But a JSON RPC request is protocol agnostic: one could use JSON RPC with websockets, or over any other protocol. Even when you're sure you need nothing but HTTP, HTTP serves as a transport protocol for JSON RPC, no HTTP semantics apply.

Being a HTTP request, FormRequest is tightly coupled to the Laravel HTTP router. It can build HTTP routes, make HTTP redirects (should not happen with JSON RPC requests, but it still can), etc.

So, FormRequest just cannot represent a JSON RPC method call. These are two different things, with different semantics. 

While it's technically possible to make a workaround by building a "simulated" HTTP request from JSON RPC method parameters, it's an ugly hack that mixes concerns and depends on Laravel internals which might change from version to version, making such an implementation extremely unreliable.

But, I totally agree that the idea of FormRequest is great, and it'd be nice to use the same approach to validation with JSON RPC.

One way we can do that is to implement our own JsonRpcFormRequest. It could have the rules() method and do the validation the same way FormRequest does, but it won't be a subclass of Illuminate\Http\Request, and error responses can be done in the JSON RPC way. In this case, the app code should look the same as with FormRequest, but, instead of extending FormRequest, JsonRpcFormRequest is extended. If due to some reason you need to share validation rules between JSON RPC requests and HTTP FormRequests, it can be done with a trait.

What do you think?

**Comment by kbaryshnikov** [comment]
Please try the `feature/form-request` branch.

See [tests/ServerFormRequestTest](https://github.com/upgate/laravel-jsonrpc/blob/feature/form-request/tests/ServerFormRequestTest.php#L152) for an example.

You can [point to a git branch](http://vvv.tobiassjosten.net/php/have-composer-use-development-branches/) on Github in composer.json.

**Comment by TheLazzziest** [comment]
I agree with that the communication over JSON RPC is agnostic. However, you build a library and that's the main difference between a standalone json-rpc server implementation and your solution which works on top of the Laravel framework. So, I think, it's reasonable to support such common features as FormRequest. Besides, you parse incoming request using the default [request](https://github.com/laravel/framework/blob/5.6/src/Illuminate/Http/Request.php) object. One way or another you can't drop it's features and start making your own, elsewise your library would lose its gist and it would become useless. 

Thinking about your implementation of FormRequest.
> One way we can do that is to implement our own JsonRpcFormRequest. It could have the rules() method and do the validation the same way FormRequest does, but it won't be a subclass of Illuminate\Http\Request, and error responses can be done in the JSON RPC way. In this case, the app code should look the same as with FormRequest, but, instead of extending FormRequest, JsonRpcFormRequest is extended. If due to some reason you need to share validation rules between JSON RPC requests and HTTP FormRequests, it can be done with a trait.

Why don't you use built-in traits which are used in default FormRequest and Request classes. There are already all required logic for managing and validating HTTP request. I mean the following:
 [Concerns\InteractsWithInput](https://github.com/laravel/framework/blob/5.6/src/Illuminate/Http/Concerns/InteractsWithInput.php),
[Macroable](https://github.com/laravel/framework/blob/5.6/src/Illuminate/Support/Traits/Macroable.php) which are used in default Request object, [ValidatesWhenResolved](https://github.com/laravel/framework/blob/5.6/src/Illuminate/Contracts/Validation/ValidatesWhenResolved.php) and [ValidatesWhenResolvedTrait](https://github.com/laravel/framework/blob/5.6/src/Illuminate/Validation/ValidatesWhenResolvedTrait.php) which are used in FormRequest class. They are self descriptive and don't depend on any components which add the required flexibility. I think all this stuff comprise minimal functionality for comfortable work with requests. 

Make a default class Requet which can be work as is or can be inherited. Make FormRequet class which inherits the Request class and add validation interface and validation trait which i mentioned above. Then you shouldn't be worry about which object is passed to controller action, you just obey users to use your request implementation and inherit from it. It will be looked something like this:
[RouteDispatcher ](https://github.com/upgate/laravel-jsonrpc/blob/master/src/Server/RouteDispatcher.php)
` if ($class) {
                try {
                    /** check for class itself or for its connection with the given class */
                    if ($class->isSubclassOf(FormRequest::class)) {
                        $args[] = $this->makeFormRequest($class->name, $requestParams);
                    } else {
                         throw new InvalidArgumentException("Class {$class} must implement " Request::class);
                    }
                } catch (ReflectionException $e) {
                    throw new InternalErrorException($e->getMessage(), 0, $e);
                }
            } else {`

**Comment by kbaryshnikov** [comment]
I do not see how it is possible to represent a JSON-RPC method call with Laravel Request.

> Besides, you parse incoming request using the default request object

No. I am not using it. There's nothing useful for JSON-RPC there.
You could be confused by this line:
```
return $jsonRpcServer->run($request);
```
Here, $request is just an opaque parameter that gets passed through middlewares. The purpose of this is to reuse existing middleware depending solely on HTTP request (like `Authenticate`). If you do not need such middleware, it's 100% ok to just call `$jsonRpcServer->run()`

As for Laravel traits and stuff... Unfortunately, the Laravel design makes things too coupled. In Laravel, it's assumed that a single HTTP request matches a single action, and, thus, we can just look at the HTTP request and say which one call we're working with. With JSON-RPC batches, this is just false.

JSON-RPC has nothing to do with HTTP. The [JSON-RPC 2.0 Specification](http://www.jsonrpc.org/specification) says:

> JSON-RPC is a stateless, light-weight remote procedure call (RPC) protocol. Primarily this specification defines several data structures and the rules around their processing. It is transport agnostic in that the concepts can be used within the same process, over sockets, over http, or in many various message passing environments.

I understand that you're looking at the simple case: there's a JSON-RPC request via HTTP, the request comes in the body, and it's a single request with one method call. In this case, yeah we could probably just think of JSON-RPC method parameters the same way as about POST data. But it's just a special case. JSON-RPC is much more powerful:

- there are batches, and batches allow for parallel processing of each method call;
- JSON-RPC can be used outside the HTTP protocol context.

Let's see two examples.

1) Imagine there's a mobile app for conductors to register passengers. The internet connection is unreliable on a moving train, so the results are sent to the server in batches once internet connection exists, in a single HTTP request:
```
[
        {"jsonrpc": "2.0", "method": "ticket.registerOnBoard", "params": {"barcode": "123"}, "id": "1"},
        {"jsonrpc": "2.0", "method": "ticket.registerOnBoard", "params": {"barcode": "456"}, "id": "2"},
       ...
]
```
How do you inject that into a single Request? What if all these calls are handled in parallel (for instance, multiplexed with a library like reactphp)?

2) Imagine a service like [CloudConvert](https://cloudconvert.com/). Say, a file gets uploaded onto a GFS cluster, and then a convert operation is queued with SQS, as a JSON-RPC call. Then, converter servers read the queue and execute the calls (with laravel queue, for example).

```
{"jsonrpc": "2.0", "id": 1, "method": "convert.video", "params": {"source": "/gfs/src/path", "target": "/gfs/target/path", "type": "video", "transcode": {...}}}
```

```
// Somewhere in cli bootstrapping...
$server->router()->bindController('convert', ConvertController::class);
// Job...
use Upgate\LaravelJsonRpc\Server\Server as JsonRpcServer;
class ConvertJob ... {
    public function handle(JsonRpcServer $server) {
        $server->setPayload($this->jsonRpcCall);
        $server->run();
    }
}
```

And this is run via php-cli, with a command like `artisan queue:work`. Completely urelated to HTTP. Nothing to do with HTTP request here at all.

**Comment by TheLazzziest** [comment]
First of all, I want to thank you for adding form request. You make my life much easier))

As for request object.

> I do not see how it is possible to represent a JSON-RPC method call with Laravel Request.

I proposed this way like an option in addition to your primary request object. Let me give you the example based on my recent task. I want to manage HTTP request over JSON-RPC protocol, including the work with headers, because I use json web token authorization [library](https://github.com/tymondesigns/jwt-auth). Now, I have to do the following in order get authorization header from request:
``` php   
public function signout()
    {
        try {
            $user = auth()->user();
            $signinRepo = $this->facade->getSigninTokenRepo();
            //@TODO:REFFACTOR
            $bearer = request()->header('Authorization');
            $bearer = explode(" ", $bearer);
            /** @var $token SigninToken */
            $token = $signinRepo->findTokenByUserIdAndToken($user->getAuthIdentifier(), $bearer[1]);
            throw_if(empty($token), ValidatorException::class,
                __("Invalid authentication token has been provided")
            );
``` 

or that 

```php   
public function signout(Request $request)
    {
        try {
            $user = auth()->user();
            $signinRepo = $this->facade->getSigninTokenRepo();
            //@TODO:REFFACTOR
            $bearer = $request->header('Authorization');
            $bearer = explode(" ", $bearer);
            /** @var $token SigninToken */
            $token = $signinRepo->findTokenByUserIdAndToken($user->getAuthIdentifier(), $bearer[1]);
            throw_if(empty($token), ValidatorException::class,
                __("Invalid authentication token has been provided")
            );
```
 
I need to do that in order to get header, and it makes the code not obvious. On then one hand, if I want to validate request, I have to use FormRequest object. On another one, if I want to manage request headers, I have to use default Request object. But if I want to do both task, I need to pass both request objects in order to be able to do that. Don't you agree that it's not the best option, do you? 
So, it's up to you how to fix this issue. But I recommend to use default request object functionality.

**Comment by TheLazzziest** [comment]
As concerns json-rpc protocol itself

> JSON-RPC has nothing to do with HTTP. The JSON-RPC 2.0 Specification says:
>>JSON-RPC is a stateless, light-weight remote procedure call (RPC) protocol. Primarily this specification defines several data structures and the rules around their processing. It is transport agnostic in that the concepts can be used within the same process, over sockets, over http, or in many various message passing environments.

Yes, I completely agree with you that it's hardly connected with the HTTP protocol. However, it's transport agnostic, as you mentioned. And it doesn't define the communication rules between the client and the server. But it's define only the applying format of the client to the server. Thus, I believe we can use it in whatever context (Http, socket and etc.). The only thing we have to do is to define requirements which we must support for particular task when we want to use json-rpc.

As for handling batch request, 

> Let's see two examples.
> Imagine there's a mobile app for conductors to register passengers. The internet connection is unreliable on a moving train, so the results are sent to the server in batches once internet connection exists, in a single HTTP request:
[
        {"jsonrpc": "2.0", "method": "ticket.registerOnBoard", "params": {"barcode": "123"}, "id": "1"},
        {"jsonrpc": "2.0", "method": "ticket.registerOnBoard", "params": {"barcode": "456"}, "id": "2"},
       ...
]
> How do you inject that into a single Request? What if all these calls are handled in parallel (for instance, multiplexed with a library like reactphp)?
Imagine a service like CloudConvert. Say, a file gets uploaded onto a GFS cluster, and then a convert operation is queued with SQS, as a JSON-RPC call. Then, converter servers read the queue and execute the calls (with laravel queue, for example).
{"jsonrpc": "2.0", "id": 1, "method": "convert.video", "params": {"source": "/gfs/src/path", "target": "/gfs/target/path", "type": "video", "transcode": {...}}}

There is no boundaries to make it. Especially, when you implement your own server

**Comment by TheLazzziest** [comment]
So, what do you think?

**Comment by kbaryshnikov** [comment]
> I need to do that in order to get header, and it makes the code not obvious. On then one hand, if I want to validate request, I have to use FormRequest object. On another one, if I want to manage request headers, I have to use default Request object. But if I want to do both task, I need to pass both request objects in order to be able to do that.

Sure, because these are two different requests.

```
// All these lines are a HTTP request
POST /jsonrpc HTTP/1.x
Authorization: Basic <base64>
Content-Type: application/json

[
{"jsonrpc": "2.0", "method": "first", "params": {...}, "id": "1"}, // First JSON-RPC request
{"jsonrpc": "2.0", "method": "second", "params": {...}, "id": "2"} // Second JSON-RPC request
]
```

Explicit is better than implicit.
Two injections are explicit, you always know what you're working with.
Patching a HTTP request object is implicit and non-obvious.

Also, if we do patch it, what's the good place for parameters? Replace POST data? But what if, for example, you want to combine JSON-RPC calls with multipart/form-data requests?

```
POST /jsonrpc HTTP/1.x
Content-Type: multipart/form-data; boundary=XXXXX

--XXXXX
Content-Disposition: form-data; name="metadata"

{...}
--XXXXX
Content-Disposition: form-data; name="file"; filename="photo.jpg"
Content-Type: image/jpeg

<binary data>
--XXXXX
Content-Disposition: form-data; name="jsonrpc"

[
{"jsonrpc": "2.0", "method": "first", "params": {...}, "id": "1"},
{"jsonrpc": "2.0", "method": "second", "params": {...}, "id": "2"}
]
```

```
if (is_multipart_request) {
    $server->setPayload($request->request->get('jsonrpc'));
}
```

What do we do? Store the original POST data, merge JSON-RPC params into, and then re-merge for every JSON-RPC call in the batch? What if there are name conflicts? Reminds me of register_globals.

**Comment by TheLazzziest** [comment]
I get what you mean.))  It takes some time for me to grasp the meaning of your words. For now, the most appropriate solution is to use two request objects: one is yours and another one is the laravel's default object for mentioned above cases. 
However, I think your request object should deal with such kind of situations to add flexibility and remove unnecessary complexity to managing json-rpc methods. Otherwise, this task is really intrigued me and I'll try to look forward for some implementations which overcome such obstacles. Anyway, thanks for your replies. They were really helpful.

**Comment by kbaryshnikov** [comment]
Anyway, it seems to be useful enough in the current state. I've just merged the `feature/form-request` branch into master and released v0.3.0.

If you come up with something better, feel free to send PRs.
